Drop IsDeprecatedWeakRefSmartPointerException for RemoteLegacyCDM

cdumez opened the pull request:
#### ee6907b1f48ccc9474e4006cb9955e625e4169dd
<pre>
Drop IsDeprecatedWeakRefSmartPointerException for RemoteLegacyCDM
<a href="https://bugs.webkit.org/show_bug.cgi?id=281791">https://bugs.webkit.org/show_bug.cgi?id=281791</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.cpp:
(WebCore::platformRegisterFactories):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivate.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateClearKey.cpp:
(WebCore::LegacyCDMPrivateClearKey::create):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateClearKey.h:
(WebCore::LegacyCDMPrivateClearKey::LegacyCDMPrivateClearKey): Deleted.
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateMediaPlayer.cpp:
(WebCore::CDMPrivateMediaPlayer::create):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateMediaPlayer.h:
(WebCore::CDMPrivateMediaPlayer::CDMPrivateMediaPlayer): Deleted.
(WebCore::CDMPrivateMediaPlayer::cdm const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.h:
(WebCore::CDMPrivateMediaSourceAVFObjC::CDMPrivateMediaSourceAVFObjC): Deleted.
(WebCore::CDMPrivateMediaSourceAVFObjC::cdm const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::create):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::initializeMockCDM):
* Source/WebCore/testing/LegacyMockCDM.cpp:
(WebCore::LegacyMockCDM::create):
* Source/WebCore/testing/LegacyMockCDM.h:
(WebCore::LegacyMockCDM::LegacyMockCDM): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setCDM):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDM.cpp:
(WebKit::RemoteLegacyCDM::create):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDM.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::registerFactory):
(WebKit::RemoteLegacyCDMFactory::createCDM):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:

Canonical link: <a href="https://commits.webkit.org/285488@main">https://commits.webkit.org/285488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4eb35f9e58d0a1eae0c2ac4936930da95d7e40f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72812 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52237 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25611 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77005 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24046 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60042 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57245 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15741 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75879 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47229 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37674 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20133 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22375 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65729 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78681 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19636 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65702 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17106 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62675 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64974 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13286 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6937 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11189 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48035 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2822 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49102 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50397 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.h:
```diff
@@ -78,7 +78,7 @@ class WEBCORE_EXPORT LegacyCDM final {
     String m_keySystem;
-    std::unique_ptr<CDMPrivateInterface> m_private;
+    RefPtr<CDMPrivateInterface> m_private;
     LegacyCDMClient* m_client;
```
Yikes a raw pointer :(

rniwa commented on Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateClearKey.h:
```diff
@@ -50,7 +48,15 @@ class LegacyCDMPrivateClearKey : public CDMPrivateInterface {
+    {
+    }
+
     LegacyCDM* m_cdm;
```
Is this raw pointer safe??

rniwa replied:
It doesn't seem like this pointer is safe given CDMPrivateInterface can now outlive LegacyCDM?

cdumez replied:
I agree. I'll fix that. Thanks for catching!

rniwa commented on Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateMediaPlayer.h:
```diff
@@ -52,7 +51,14 @@ class CDMPrivateMediaPlayer : public CDMPrivateInterface {
+        : m_cdm(cdm)
+    { }
+
     LegacyCDM* m_cdm;
```
Ditto.

rniwa commented on Source/WebCore/testing/LegacyMockCDM.h:
```diff
@@ -50,7 +49,14 @@ class LegacyMockCDM : public CDMPrivateInterface {
+        : m_cdm(cdm)
+    { }
+
     LegacyCDM* m_cdm;
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
```diff
@@ -51,20 +51,28 @@ RemoteLegacyCDMFactory::RemoteLegacyCDMFactory(WebProcess& webProcess)
-            if (weakThis)
-                return weakThis->createCDM(privateCDM);
-            return nullptr;
+        [protectedThis = Ref { *this }] (LegacyCDM& privateCDM) -> std::unique_ptr<WebCore::CDMPrivateInterface> {
```
I think this would result in memory leak, no? `registerCDMFactory` will add it to `installedCDMFactories()`, which is a `static NeverDestroyed`.

cdumez replied:
it's owned by the WebProcess. It's a singleton

rniwa replied:
ah, ok.

cdumez committed:
Drop IsDeprecatedWeakRefSmartPointerException for RemoteLegacyCDM
https://bugs.webkit.org/show_bug.cgi?id=281791

Reviewed by Ryosuke Niwa.

* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.cpp:
(WebCore::platformRegisterFactories):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDM.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivate.h:
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateClearKey.cpp:
(WebCore::LegacyCDMPrivateClearKey::create):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateClearKey.h:
(WebCore::LegacyCDMPrivateClearKey::LegacyCDMPrivateClearKey): Deleted.
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateMediaPlayer.cpp:
(WebCore::CDMPrivateMediaPlayer::create):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMPrivateMediaPlayer.h:
(WebCore::CDMPrivateMediaPlayer::CDMPrivateMediaPlayer): Deleted.
(WebCore::CDMPrivateMediaPlayer::cdm const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.h:
(WebCore::CDMPrivateMediaSourceAVFObjC::CDMPrivateMediaSourceAVFObjC): Deleted.
(WebCore::CDMPrivateMediaSourceAVFObjC::cdm const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::create):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::initializeMockCDM):
* Source/WebCore/testing/LegacyMockCDM.cpp:
(WebCore::LegacyMockCDM::create):
* Source/WebCore/testing/LegacyMockCDM.h:
(WebCore::LegacyMockCDM::LegacyMockCDM): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setCDM):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDM.cpp:
(WebKit::RemoteLegacyCDM::create):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDM.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::registerFactory):
(WebKit::RemoteLegacyCDMFactory::createCDM):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:

Canonical link: https://commits.webkit.org/285488@main
